File: src/core/rxdb/src/plugins/replication_webrtc/v1_5_status.rs
//! V1.5 Query-Demand-Loading status surface (Rust side).
//!
//! Mirrors `src/apps/business-os/rxdb/src/v1_5_status.mjs`. Filled
//! progressively by later waves; this module currently exposes the constant
//! capability name and the field list that the Rust peer must report.

pub const V1_5_QUERY_FETCH_CAPABILITY: &str = "ctox-rxdb-query-fetch-v1";

pub const V1_5_STATUS_FIELDS: &[&str] = &[
    "rxdbRuntime",
    "rxdbProtocolVersion",
    "transport",
    "peerConnected",
    "peerCapabilityQueryFetchV1",
    "queryDemandLoadingEnabled",
    "queryDemandLoadingActive",
    "queryFetchInFlight",
    "queryFetchSuccessCount",
    "queryFetchErrorCount",
    "queryFetchDedupHitCount",
    "indexedDbWorkingSetBytes",
    "indexedDbEvictionCount",
    "pinnedDocCount",
    "pinnedBytes",
    "lastQueryFetchMs",
    "lastTransportBackpressureMs",
    "lastReloadHydrationMs",
];

#[derive(Debug, Clone)]
pub struct V1_5StatusBaseline;

impl V1_5StatusBaseline {
    pub fn protocol_version(&self) -> &'static str {
        "1"
    }

    pub fn query_demand_loading_active(&self) -> bool {
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn capability_constant_is_stable() {
        assert_eq!(V1_5_QUERY_FETCH_CAPABILITY, "ctox-rxdb-query-fetch-v1");
    }

    #[test]
    fn status_field_list_contains_negotiation_fields() {
        assert!(V1_5_STATUS_FIELDS.contains(&"peerCapabilityQueryFetchV1"));
        assert!(V1_5_STATUS_FIELDS.contains(&"queryDemandLoadingActive"));
    }

    #[test]
    fn baseline_reports_protocol_version_one() {
        let baseline = V1_5StatusBaseline;
        assert_eq!(baseline.protocol_version(), "1");
        assert!(!baseline.query_demand_loading_active());
    }
}

File: src/apps/business-os/rxdb/src/v1_5_status.mjs
// V1.5 status surface. Filled progressively by later waves.
// Values reflect the V1-only baseline until Wave 1 lights up the capability
// and Wave 3 wires demand-loading into the query path. Do NOT widen the
// schema without bumping the field list documented in docs/rxdb_on-demand-load.md.

import { CTOX_QUERY_FETCH_CAPABILITY, CTOX_QUERY_RPC } from './protocol-contract.generated.mjs';

export const V1_5_QUERY_FETCH_CAPABILITY = CTOX_QUERY_FETCH_CAPABILITY;
export const V1_5_QUERY_RPC = CTOX_QUERY_RPC;

export const V1_5_STATUS_FIELDS = Object.freeze([
  'rxdbRuntime',
  'rxdbProtocolVersion',
  'transport',
  'peerConnected',
  'peerCapabilityQueryFetchV1',
  'queryDemandLoadingEnabled',
  'queryDemandLoadingActive',
  'queryFetchInFlight',
  'pendingQueryFetchCollectors',
  'queuedQueryFetchRequests',
  'maxPendingQueryFetchCollectors',
  'queryFetchSuccessCount',
  'queryFetchErrorCount',
  'queryFetchDedupHitCount',
  'indexedDbWorkingSetBytes',
  'indexedDbEvictionCount',
  'pinnedDocCount',
  'pinnedBytes',
  'lastQueryFetchMs',
  'lastTransportBackpressureMs',
  'lastReloadHydrationMs',
  'activeFileStreams',
  'pendingFileFetchCollectors',
  'maxPendingFileFetchCollectors',
  'fileBytesReceived',
  'fileStreamErrors',
  'fileStreamDedupHits',
  'lastFileFetchMs',
  'localPushChangedSinceCalls',
  'localPushChangedSinceScannedRows',
  'localPushChangedSinceScanLimitHits',
  'localPushChangedSinceMaxScannedRows',
  'clockSkewDetected',
  'nativeClockOffsetMs',
  'nativeClockObservedAtMs',
  'code',
]);

export function createV1_5StatusState() {
  return {
    rxdbRuntime: 'ctox-rxdb-js',
    rxdbProtocolVersion: '1',
    transport: 'webrtc',
    peerConnected: false,
    peerCapabilityQueryFetchV1: false,
    queryDemandLoadingEnabled: false,
    queryDemandLoadingActive: false,
    queryFetchInFlight: 0,
    pendingQueryFetchCollectors: 0,
    queuedQueryFetchRequests: 0,
    maxPendingQueryFetchCollectors: 0,
    queryFetchSuccessCount: 0,
    queryFetchErrorCount: 0,
    queryFetchDedupHitCount: 0,
    indexedDbWorkingSetBytes: 0,
    indexedDbEvictionCount: 0,
    pinnedDocCount: 0,
    pinnedBytes: 0,
    lastQueryFetchMs: null,
    lastTransportBackpressureMs: null,
    lastReloadHydrationMs: null,
    activeFileStreams: 0,
    pendingFileFetchCollectors: 0,
    maxPendingFileFetchCollectors: 0,
    fileBytesReceived: 0,
    fileStreamErrors: 0,
    fileStreamDedupHits: 0,
    lastFileFetchMs: null,
    localPushChangedSinceCalls: 0,
    localPushChangedSinceScannedRows: 0,
    localPushChangedSinceScanLimitHits: 0,
    localPushChangedSinceMaxScannedRows: 0,
    clockSkewDetected: false,
    nativeClockOffsetMs: 0,
    nativeClockObservedAtMs: null,
    code: null,
  };
}

export function projectStatusFromSidecar(state, sidecarStats, registry = null) {
  const next = { ...state };
  if (sidecarStats) {
    next.indexedDbWorkingSetBytes = sidecarStats.estimatedBytes || 0;
  }
  if (registry?.pinnedDocCount !== undefined) next.pinnedDocCount = registry.pinnedDocCount;
  if (registry?.pinnedBytes !== undefined) next.pinnedBytes = registry.pinnedBytes;
  return next;
}

export function snapshotV1_5Status(state) {
  const snapshot = {};
  for (const field of V1_5_STATUS_FIELDS) {
    snapshot[field] = state?.[field] ?? null;
  }
  return snapshot;
}
